File: src/main/java/xyz/devvydont/smprpg/items/blueprints/block/rawmaterials/SulfurBlock.java
package xyz.devvydont.smprpg.items.blueprints.block.rawmaterials;

import org.bukkit.NamespacedKey;
import org.bukkit.inventory.CraftingRecipe;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.ShapedRecipe;
import org.bukkit.inventory.recipe.CraftingBookCategory;
import xyz.devvydont.smprpg.SMPRPG;
import xyz.devvydont.smprpg.block.CustomBlock;
import xyz.devvydont.smprpg.items.CustomItemType;
import xyz.devvydont.smprpg.items.blueprints.block.BlockBlueprint;
import xyz.devvydont.smprpg.items.interfaces.ICraftable;
import xyz.devvydont.smprpg.items.interfaces.ISellable;
import xyz.devvydont.smprpg.services.ItemService;

import java.util.Collection;
import java.util.List;

public class SulfurBlock extends BlockBlueprint implements ICraftable, ISellable {

    public SulfurBlock(ItemService itemService, CustomItemType type) {
        super(itemService, type);
    }

    @Override
    public CustomBlock getCustomBlock() {
        return CustomBlock.SULFUR_BLOCK;
    }

    @Override
    public NamespacedKey getRecipeKey() {
        return new NamespacedKey(SMPRPG.getPlugin(), this.getCustomItemType().getKey() + "_recipe");
    }

    @Override
    public CraftingRecipe getCustomRecipe() {
        var recipe = new ShapedRecipe(this.getRecipeKey(), generate());
        recipe.shape(
                "sss",
                "sss",
                "sss"
        );
        recipe.setIngredient('s', ItemService.generate(CustomItemType.SULFUR));
        recipe.setCategory(CraftingBookCategory.MISC);
        return recipe;
    }

    @Override
    public Collection<ItemStack> unlockedBy() {
        return List.of(
                itemService.getCustomItem(CustomItemType.SULFUR)
        );
    }

    @Override
    public int getWorth(ItemStack item) {
        return (CustomItemType.SULFUR.Worth * 9) * item.getAmount();
    }
}

File: src/main/java/xyz/devvydont/smprpg/items/blueprints/block/rawmaterials/SulfurBlock.kt
package xyz.devvydont.smprpg.items.blueprints.block.rawmaterials

import org.bukkit.NamespacedKey
import org.bukkit.inventory.CraftingRecipe
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.ShapedRecipe
import org.bukkit.inventory.recipe.CraftingBookCategory
import xyz.devvydont.smprpg.SMPRPG.Companion.plugin
import xyz.devvydont.smprpg.block.CustomBlock
import xyz.devvydont.smprpg.items.CustomItemType
import xyz.devvydont.smprpg.items.blueprints.block.BlockBlueprint
import xyz.devvydont.smprpg.items.interfaces.ICraftable
import xyz.devvydont.smprpg.items.interfaces.ISellable
import xyz.devvydont.smprpg.items.interfaces.ICompressible
import xyz.devvydont.smprpg.items.interfaces.ICompressible.CompressionStep
import xyz.devvydont.smprpg.services.ItemService
import xyz.devvydont.smprpg.services.ItemService.Companion.blueprint
import xyz.devvydont.smprpg.services.ItemService.Companion.generate

class SulfurBlock(itemService: ItemService, type: CustomItemType) : BlockBlueprint(itemService, type),
    ISellable, ICompressible {
    override fun getCustomBlock(): CustomBlock {
        return CustomBlock.SULFUR_BLOCK
    }

    override fun getWorth(item: ItemStack): Int {
        val dust = itemService.getCustomItem(CustomItemType.SULFUR)
        val bp = blueprint(dust)
        if (bp is ISellable) {
            return ((bp as ISellable).getWorth(dust) * 9) * item.amount
        }
        return 0
    }

    override val decompressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.SULFUR_BLOCK ->
            CompressionStep(itemService.getBlueprint(CustomItemType.SULFUR) as ICompressible, 1, 9)
        else -> null
    }

    override val compressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.SULFUR_BLOCK ->
            CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_SULFUR) as ICompressible, 9, 1)
        else -> null
    }
}
